### ISSUE TITLE
[RFC] Quantization Framework For Disk Optimized Vector Search

### ISSUE THREAD

**Issue by Vikasht34** [opensearch-project/k-NN#1889]
# Introduction
This document proposes a general-purpose framework for vector quantization. This framework aims to utilize various quantization techniques, including product quantization, binary quantization, and scalar quantization. The goal is to compress floating-point vectors into bytes or bits, significantly reducing memory usage for storing and processing vector data in semantic search applications.

---

## Requirements

| Requirement ID | Description
| -------------- | ----------- |
| R-1            | Effective Binary Quantizer for both float and byte vectors | 
| R-2            | Extensible framework for various quantization techniques | 
| R-3            | Memory-efficient handling of large datasets |
| R-4            | Extensible for future quantization techniques and integrations |
| R-5            | Client-selectable quantization techniques based on needs |
| R-6            | Implement sampler for training data |

---

## Success Metrics

| Metric ID | Description |
| --------- | ----------- |
| M-1       | Quantization latency should be less than 20ms |
| M-2       | Memory usage of the framework (dependent on vector dimension and quantization type) |

---

## State of the System

Currently, we have the following quantization techniques:
1. **Product Quantization**: Requires an additional training step.
2. **Byte Quantization**: Implemented through Lucene.
3. **Scalar Quantization**: Implemented through FAISS (FP16).

---

### Responsibilities of Quantizer Framework

1. **Support Multiple Quantization Types**:
   - Supports space-compression (e.g., PQ) and value-compression (e.g., Scalar Quantization) techniques.
   - Extensible for future methods like PCA or 2-bit quantization.
   - New methods can be added with minimal changes to existing code.

2. **Clear Interface for Quantization and Training**:
   - Clients can specify and configure parameters for different techniques.
   - Data sources for quantization can be specified.

3. **Efficiency**:
   - Optimize memory usage during the quantization process.

4. **Quantization Framework will sample records from training.**

5. **Stateless**:
   - Does not store quantization state or thresholds.

6. **Metrics Tracking**:
   - Time taken during training and quantization.
   - Memory consumption.

7. **Decision Making**:
   - The framework does not decide when to re-quantize; it accepts quantization states.

### Components of Quantization Framework

1. **Quantization Orchestrator**
2. **Sampler**
3. **Quantizers**
![Untitled Diagram-Page-7 (7) (1)](https://github.com/user-attachments/assets/4a71004d-558d-40e6-9ffb-8f140f32842b)

#### Quantization Orchestrator Responsibilities

- Manages the quantization process.
- Identifies appropriate quantizers based on parameters.
- Calls individual quantizers for training.
- Returns quantization state and quantizer instances to the client.
- Emits metrics on quantization and training performance.

#### Quantization Build Param
The QuantizationParams class encapsulates the parameters necessary identifying end quantizer based on multiple parameters. 

* **Quantization Type** :- 
    * Space_Quantization :- could be seen as compressing data in a dimensional or spatial sense, like Product Quantization which reduces dimensions of data.
    * Value_Quantization :- refers to compressing non-dimensional aspects, such as scalar values, like Scalar Quantization.

- **Space Quantization Types**: PQ, etc.
- **Value Quantization Types**: SQ
- **SQ Types**: FP16, INT8, INT6, INT4, ONE_BIT, TWO_BIT

#### Vector access during training and quantization
As our one of the main requirement quantization framework is memory efficient , which means if I have to quantize 1M vectors of dimension D , it’s can not be loaded into memory for quantization. Based on below requirements , interfaces will be defined. 

* **Training** :- 
    * During training , Quantization Orchestrator will use sampler to sample records. 
    * Uses an iterator-based interface to get vector from client for memory efficiency.
    * Returns a QuantizationState, a class representing the state which is produced by a quantizer after training, including thresholds for quantizing vectors , quantization parameters, etc. Quantization State is helpful during merging segments and during quantizing query where qunatization function depends on some scalar thresholds to quantize the query. 
* **Quantization** :- 
    * Quantization Orchestrator provides a quantizer based on the QuantizationState/QuantizationParams.
    * The client uses the quantizer to quantize vector.
    * Returns an object with the quantized vector and relevant parameters.
 
#### Sampler
Responsibility of Sampler is to sample the records for training. It can use any of the sampling techniques like Reservoir Sampling, Snowball Sampling , Quota Sampling, Simple Random sampling , Stratified Sampling to sample the records.  For now we will use Reservoir Sampling, and in future we find better sampler then it will be replaced.  
![Untitled Diagram-Page-12 (1)](https://github.com/user-attachments/assets/f963ee9d-32d7-4d93-ae8d-8a0c6bc28357)

#### Quantizer

1. Product Quantizer :- A Product Quantizer (PQ) is  technique used in data compression and similarity search, particularly for high-dimensional vectors. It works by decomposing a high-dimensional space into a Cartesian product of lower-dimensional subspaces, which are then quantized separately. Each subspace is assigned a codebook, and the combination of these codebooks forms the quantized representation of the original data. 
2. Scalar Quantizer:- A scalar quantizer is a fundamental component  data compression. It's a technique used to reduce the precision of a continuous-valued signal by mapping each sample to a smaller set of discrete values, called quantization levels.
3. Binary Quantization :-  Binary quantization is special type of scalar quantization , unlike scalar quantizers, which map continuous values to a set of discrete levels, a binary quantizer reduces the precision of a signal to just two levels: 0 or 1. This results in a binary representation of the signal, where each sample is represented by a single bit.

---

#### Indexing Flow

* **End to end data flow  for Quantizers which  requires on the fly training and graph creation is not incremental**:-

This diagram shows how our Quantizer Framework integrates with Lucene's Scalar Quantization and to demonstrate the flexibility of the framework.
![Untitled Diagram-Page-8 (4) (1)](https://github.com/user-attachments/assets/6235f545-6f2a-4f25-8ee0-fa6e85dde9ae)

In the diagram, you can see that we perform training during the finish function of the FieldWriter and quantization during the flush function of the writer. However, we have the flexibility to perform both training and quantization during the flush function.
The major advantage of performing training in the finish function is the ability to separate the storage of quantized vectors and raw vectors into different files easily. 

* **End to end data flow  for Binary Quantization function like Sign Magnitude which does not require on the fly training and Graph creation is not incremental. **:-

This diagram show integration of Binary Quantizers function like Sign Magnitude which does not require training and graph construction using FAISS Index Binary
![Untitled Diagram-Page-9 (2) (1)](https://github.com/user-attachments/assets/474c0a1b-1af5-46a3-8d08-fb7271c8751f)

From the above diagram ,  Quantization Orchestrator is getting initialised in field writer, though we have option of initialising on flush method. initialised in field writer will give us flexibility to have separate quantization at field info level , which is good of isolation. 

* **End to end data flow  for Binary Quantization function like Sign Magnitude which does not require on the fly training and Graph creation is  incremental. . **:-

![Untitled Diagram-Page-10 (1)](https://github.com/user-attachments/assets/53fd2ad2-6c48-4613-bb61-a74b82c7678c)

#### Query Flow
* **Query Flow when Query needs to be quantized using scalar value which was computed during training**:-
![Untitled Diagram-Page-11 (1)](https://github.com/user-attachments/assets/a3ff0ae4-2de0-4d2e-979c-fabc00e5821c)
During the query flow, if training is involved in the quantization process, the quantization state will be saved per segment. This state includes not only the parameters used during quantization but also the training results. These are essential for accurately quantizing the vector.When a query is processed, the quantization state is retrieved and passed to the quantization orchestrator along with the vector and its destination. The quantization orchestrator will then identify and call the appropriate quantizer to handle the query.

* **Query Flow when Query needs to be quantized using static quantization function which does not require training without saving Quantization State to Segment.**:-
![Untitled Diagram-Page-13 (1)](https://github.com/user-attachments/assets/e5ab9f92-1f50-4575-a23e-011108a467b3)

In this query flow, we do not store the quantization state in the segment. Instead, we construct quantization build parameters from field information attributes. The quantization framework then identifies the appropriate quantizer based on these parameters and quantizes the vector.

---
